lms/product/canvas/_plugin/course_copy.py

```diff
@@ -29,19 +24,17 @@ def find_matching_file_in_course(self, course_id, file_ids):
 
         Return None if no matching file is found.
         """
+        # Find the files in the current course. We call this for the side effect of storing the files in the DB
+        _ = self._api.list_files(course_id)
+
         for file_id in file_ids:
             file = self._file_service.get(file_id, type_="canvas_file")
 
             if not file:
                 continue
 
-            for file_dict in self._api.list_files(course_id):
-                if (
-                    file_dict["display_name"] == file.name
-                    and file_dict["size"] == file.size
-                    and str(file_dict["id"]) != file.lms_id
-                ):
-                    return str(file_dict["id"])
+            if new_file := self._file_service.find_copied_file(course_id, file):
```

Comment:
This method already check for size name and id:

https://github.com/hypothesis/lms/blob/380f760b543c60750f81614c9dad5620471744e6/lms/services/file.py#L23

Reply:
Something seems a bit iffy to the way the file service and the plugins + helpers interact.

Our general pattern elsewhere was that plugins are called from services, and general consumers don't need to know about their existence to get the job done.

So:

 * General code > service > plugin

In this case I think we have a dependency inversion where we have:

 * General code > plugin -> service
 
I would like to see what a version where you ask the file service for what you want and it "does the right" thing might look like. At the moment a consumer of the file service will need to be quite informed about the correct usage of the copy plugin to be able to get the correct answer.